File: ios/App/App/TriggerTelemetry.swift
import Foundation
import CoreLocation

/**
 TriggerTelemetry — banco di misura per tarare il geofencing predittivo.

 PORT di android/.../geofence/TriggerTelemetry.kt (stesso header CSV, così i
 file delle due piattaforme si analizzano con lo stesso foglio di calcolo).

 PERCHÉ ESISTE
 -------------
 `PredictiveTrigger.tLead*` e `corridor*` sono valori di partenza scelti a
 tavolino, non taratura: dipendono dalla latenza reale del fix sul
 dispositivo, dalla densità urbana e dall'andatura dell'utente. Senza misura,
 cambiarli è indovinare — e un annuncio troppo anticipato è percepito come
 falso positivo esattamente come uno in ritardo.

 COME SI LEGGE
 -------------
 Il file sta in Documents, quindi è estraibile senza Xcode:
   - 3uTools → File Manager → app WIP → Documents → wip_trigger_telemetry.csv
   - oppure Finder/iTunes se l'app espone UIFileSharingEnabled
 In tempo reale, con l'app collegata: i print compaiono nella console.

 COME SI TARA
 ------------
   - FIRE sistematicamente troppo vicino  → alzare tLead
   - FIRE su POI di lato / mai visitati   → abbassare corridor
   - molti `fail-open`                    → il collo di bottiglia è
     l'accuratezza GPS, non le costanti: agire sul campionamento.

 Scrittura best-effort: la telemetria non deve mai poter rompere un trigger.
 */
enum TriggerTelemetry {

    private static let fileName = "wip_trigger_telemetry.csv"

    /// Tetto dimensione file: oltre, si riparte da capo.
    private static let maxBytes = 512 * 1024

    private static let header =
        "timestamp,poi_id,poi_name,phase,decision,reason,used_prediction," +
        "t_cpa_s,d_cpa_m,d_now_m,radius_m,fix_age_ms,accuracy_m,speed_ms,bearing_deg,mode\n"

    /// Interruttore: spegnibile da UserDefaults senza ricompilare.
    private static var isEnabled: Bool {
        UserDefaults.standard.object(forKey: "telemetryEnabled") as? Bool ?? true
    }

    private static let isoFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private static var fileURL: URL? {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
            .first?.appendingPathComponent(fileName)
    }

    static func log(
        poiId: String,
        poiName: String,
        phase: String,
        result: PredictiveTrigger.Result,
        location: CLLocation,
        isDriving: Bool,
        radiusM: Double
    ) {
        let fixAgeMs = max(0, Date().timeIntervalSince1970 * 1000
                              - location.timestamp.timeIntervalSince1970 * 1000)
        let accuracy = location.horizontalAccuracy
        let speed = location.speed
        let course = location.course

        // Console sempre: è il canale utile mentre si cammina col telefono in mano.
        print("[WipTelemetry] \(phase) \(result.decision) \"\(poiName)\" " +
              "t_cpa=\(num(result.tCpaSeconds))s d_cpa=\(num(result.dCpaMeters))m " +
              "d_now=\(Int(result.distanceNowMeters))m r=\(Int(radiusM))m " +
              "age=\(Int(fixAgeMs))ms acc=\(Int(accuracy))m v=\(num(speed))m/s " +
              "brg=\(Int(course))° pred=\(result.usedPrediction) — \(result.reason)")

        guard isEnabled, let url = fileURL else { return }

        let row = [
            isoFormatter.string(from: Date()),
            poiId,
            csv(poiName),
            phase,
            String(describing: result.decision),
            csv(result.reason),
            String(result.usedPrediction),
            num(result.tCpaSeconds),
            num(result.dCpaMeters),
            String(Int(result.distanceNowMeters)),
            String(Int(radiusM)),
            String(Int(fixAgeMs)),
            String(Int(accuracy)),
            num(speed),
            String(Int(course)),
            isDriving ? "driving" : "walking"
        ].joined(separator: ",") + "\n"

        do {
            let fm = FileManager.default
            let size = (try? fm.attributesOfItem(atPath: url.path)[.size] as? Int) ?? 0
            if !fm.fileExists(atPath: url.path) || (size ?? 0) > maxBytes {
                try header.write(to: url, atomically: true, encoding: .utf8)
            }
            if let handle = try? FileHandle(forWritingTo: url) {
                handle.seekToEndOfFile()
                if let data = row.data(using: .utf8) { handle.write(data) }
                try? handle.close()
            }
        } catch {
            // Best-effort: mai far fallire un trigger per la telemetria.
            print("[WipTelemetry] write failed: \(error.localizedDescription)")
        }
    }

    static func clear() {
        guard let url = fileURL else { return }
        try? header.write(to: url, atomically: true, encoding: .utf8)
    }

    private static func num(_ v: Double) -> String {
        if v.isNaN || v.isInfinite || v < -9_000_000 { return "" }
        return String(format: "%.1f", v)
    }

    /// Virgole e virgolette romperebbero il CSV.
    private static func csv(_ s: String) -> String {
        if s.contains(",") || s.contains("\"") {
            return "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\""
        }
        return s
    }
}

File: android/app/src/main/java/com/itaintasca/app/geofence/TriggerTelemetry.kt
package com.itaintasca.app.geofence

import android.content.Context
import android.location.Location
import android.util.Log
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * TriggerTelemetry — banco di misura per tarare il geofencing predittivo.
 *
 * PERCHÉ ESISTE
 * -------------
 * `PredictiveTrigger.T_LEAD_*` e `CORRIDOR_*` sono valori di partenza scelti
 * a tavolino. Non sono taratura: dipendono dalla latenza reale del fix sul
 * dispositivo, dalla densità urbana e dall'andatura dell'utente.
 *
 * Senza misura, cambiare quelle costanti è indovinare — e un annuncio troppo
 * anticipato è percepito come falso positivo esattamente come uno in ritardo.
 *
 * COSA REGISTRA
 * -------------
 * Una riga CSV per ogni valutazione: età del fix, accuratezza, velocità,
 * t_cpa e d_cpa previsti, distanza al momento della decisione, esito.
 *
 * COME SI LEGGE (senza root, da PC):
 *   adb shell run-as com.itaintasca.app cat files/wip_trigger_telemetry.csv > telemetria.csv
 * oppure via logcat in tempo reale mentre si cammina:
 *   adb logcat -s WipTelemetry
 *
 * COME SI TARA
 * ------------
 * Dopo alcune camminate, per ogni POI confrontare `d_now` al momento del FIRE
 * con la distanza minima realmente raggiunta poi (`d_cpa` è la previsione).
 *   - FIRE sistematicamente troppo vicino → alzare T_LEAD
 *   - FIRE su POI mai visitati / di lato → abbassare CORRIDOR
 *   - molti `fail-open` → il collo di bottiglia è l'accuratezza GPS,
 *     non le costanti: intervenire sul campionamento, non sulle soglie.
 *
 * Scrittura best-effort: qualunque errore di I/O viene ignorato. La
 * telemetria non deve mai poter rompere un trigger.
 */
object TriggerTelemetry {

    private const val TAG = "WipTelemetry"
    private const val FILE_NAME = "wip_trigger_telemetry.csv"

    /** Tetto dimensione file: oltre, si riparte da capo (nessuna rotazione multipla). */
    private const val MAX_BYTES = 512 * 1024

    private const val HEADER =
        "timestamp,poi_id,poi_name,phase,decision,reason,used_prediction," +
        "t_cpa_s,d_cpa_m,d_now_m,radius_m,fix_age_ms,accuracy_m,speed_ms,bearing_deg,mode\n"

    /**
     * (23/08/2026) SimpleDateFormat NON e' thread-safe, e qui scrivevano il
     * loop predittivo del servizio, il receiver dei geofence e il thread
     * dell'arrivo: due format() sovrapposti producono timestamp corrotti
     * (millisecondi di una riga dentro l'orario di un'altra), cioe' proprio la
     * colonna su cui si tara il geofencing. Uno per thread, stesso formato.
     */
    private val fmt = object : ThreadLocal<SimpleDateFormat>() {
        override fun initialValue() = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US)
    }

    /**
     * Interruttore: si può spegnere (e ACCENDERE) da SharedPreferences senza
     * ricompilare — e' cosi' che si arma la prova in strada.
     *
     * (23/08/2026) Il DEFAULT ora e' BuildConfig.DEBUG: in una build di
     * release il file CSV non si scrive finche' qualcuno non mette
     * `telemetryEnabled=true` nelle prefs. Prima era acceso per tutti, e
     * significava aprire, scrivere, svuotare e chiudere un file su ogni
     * decisione di trigger, sul thread che decide se far partire la voce.
     */
    private fun isEnabled(context: Context): Boolean =
        context.getSharedPreferences("ItaintaPrefs", Context.MODE_PRIVATE)
            .getBoolean("telemetryEnabled", com.itaintasca.app.BuildConfig.DEBUG)

    // ── Scrittura a lotti, fuori dal thread che decide ───────────────────
    // Prima ogni riga costava: File(), exists(), length() (due stat() sul
    // filesystem) e poi un appendText che apre/scrive/flusha/chiude. Ora le
    // righe si accumulano in memoria e scendono su disco a lotti, su un
    // thread suo. Il contenuto del CSV e' identico, riga per riga.
    private const val RIGHE_PER_LOTTO = 20
    private const val ATTESA_MASSIMA_MS = 10_000L

    private val righe = ArrayList<String>()
    private var ultimoFlushAt = 0L

    private val scrittore: java.util.concurrent.ExecutorService by lazy {
        java.util.concurrent.Executors.newSingleThreadExecutor { r ->
            Thread(r, "wip-telemetry").apply { isDaemon = true }
        }
    }

    private fun accoda(context: Context, riga: String) {
        val appContext = context.applicationContext
        val lotto: List<String>?
        synchronized(righe) {
            righe.add(riga)
            val adesso = System.currentTimeMillis()
            if (ultimoFlushAt == 0L) ultimoFlushAt = adesso
            lotto = if (righe.size >= RIGHE_PER_LOTTO || adesso - ultimoFlushAt >= ATTESA_MASSIMA_MS) {
                ultimoFlushAt = adesso
                ArrayList(righe).also { righe.clear() }
            } else null
        }
        if (lotto != null) scrittore.execute { scriviLotto(appContext, lotto) }
    }

    private fun scriviLotto(context: Context, lotto: List<String>) {
        if (lotto.isEmpty()) return
        try {
            val file = File(context.filesDir, FILE_NAME)
            if (!file.exists() || file.length() > MAX_BYTES) {
                file.writeText(HEADER)
            }
            file.appendText(lotto.joinToString(""))
        } catch (e: Exception) {
            // Best-effort: la telemetria non deve mai rompere un trigger.
            Log.w(TAG, "Telemetry write failed: ${e.message}")
        }
    }

    /**
     * Porta su disco tutto quello che e' ancora in memoria. Va chiamata prima
     * di LEGGERE il file (fileFor), altrimenti l'ultima manciata di righe —
     * quelle appena camminate — non c'e' ancora.
     */
    fun flush(context: Context) {
        val appContext = context.applicationContext
        val lotto: List<String>
        synchronized(righe) {
            if (righe.isEmpty()) return
            ultimoFlushAt = System.currentTimeMillis()
            lotto = ArrayList(righe)
            righe.clear()
        }
        scriviLotto(appContext, lotto)
    }

    fun log(
        context: Context,
        poiId: String,
        poiName: String,
        phase: String,
        result: PredictiveTrigger.Result,
        location: Location,
        isDriving: Boolean,
        radiusM: Float
    ) {
        val fixAgeMs = (System.currentTimeMillis() - location.time).coerceAtLeast(0L)
        val accuracy = if (location.hasAccuracy()) location.accuracy else -1f
        val speed = if (location.hasSpeed()) location.speed else -1f
        val bearing = if (location.hasBearing()) location.bearing else -1f

        // Logcat sempre: è il canale utile mentre si cammina col telefono in mano.
        Log.i(
            TAG,
            "$phase ${result.decision} \"$poiName\" " +
                "t_cpa=${fmtNum(result.tCpaSeconds)}s d_cpa=${fmtNum(result.dCpaMeters)}m " +
                "d_now=${result.distanceNowMeters.toInt()}m r=${radiusM.toInt()}m " +
                "age=${fixAgeMs}ms acc=${accuracy.toInt()}m v=${fmtNum(speed.toDouble())}m/s " +
                "brg=${bearing.toInt()}° pred=${result.usedPrediction} — ${result.reason}"
        )

        if (!isEnabled(context)) return

        try {
            accoda(
                context,
                listOf(
                    (fmt.get() ?: SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US)).format(Date()),
                    poiId,
                    csv(poiName),
                    phase,
                    result.decision.name,
                    csv(result.reason),
                    result.usedPrediction.toString(),
                    fmtNum(result.tCpaSeconds),
                    fmtNum(result.dCpaMeters),
                    result.distanceNowMeters.toInt().toString(),
                    radiusM.toInt().toString(),
                    fixAgeMs.toString(),
                    accuracy.toInt().toString(),
                    fmtNum(speed.toDouble()),
                    bearing.toInt().toString(),
                    if (isDriving) "driving" else "walking"
                ).joinToString(",") + "\n"
            )
        } catch (e: Exception) {
            // Best-effort: la telemetria non deve mai rompere un trigger.
            Log.w(TAG, "Telemetry write failed: ${e.message}")
        }
    }

    /**
     * Percorso del file, per esporlo al plugin Capacitor / condivisione.
     * Svuota prima il buffer: chi chiede il file lo vuole COMPLETO.
     */
    fun fileFor(context: Context): File {
        flush(context)
        return File(context.filesDir, FILE_NAME)
    }

    fun clear(context: Context) {
        synchronized(righe) { righe.clear() }
        try { File(context.filesDir, FILE_NAME).writeText(HEADER) } catch (_: Exception) {}
    }

    private fun fmtNum(v: Double): String =
        if (v.isNaN() || v.isInfinite()) "" else String.format(Locale.US, "%.1f", v)

    /** Virgole e virgolette romperebbero il CSV. */
    private fun csv(s: String): String =
        if (s.contains(',') || s.contains('"')) "\"" + s.replace("\"", "\"\"") + "\"" else s
}
